lang/src/org/partiql/lang/syntax/SqlParser.kt

```diff
@@ -3153,6 +3283,283 @@ class SqlParser(
+                    "<" -> {
+                        rem = rem.tail
+                        if (rem.head?.type == TokenType.OPERATOR) {
+                            when (rem.head!!.keywordText) {
+                                "-" -> EdgeType(left = true, right = false, undirected = false)
+                                "~" -> EdgeType(left = true, right = false, undirected = true)
+                                else -> errorEdgeParse()
+                            }
+                        } else {
+                            errorEdgeParse()
+                        }
+                    }
+                    "-" -> EdgeType(left = false, right = true, undirected = false)
+                    "~" -> EdgeType(left = false, right = false, undirected = true)
+                    else -> errorEdgeParse()
+                }
+            } else {
+                errorEdgeParse()
+            }
+            rem = rem.tail
+
+            return direction
+        }
+
+        fun parseRightEdgePattern(): EdgeType {
+            val direction = if (rem.head?.type == TokenType.OPERATOR) {
+                when (rem.head!!.keywordText) {
+                    "-" -> {
+                        if (rem.tail.head?.type == TokenType.OPERATOR && rem.tail.head?.keywordText == ">") {
+                            rem = rem.tail
+                            EdgeType(left = false, right = true, undirected = false)
+                        } else {
+                            EdgeType(left = true, right = false, undirected = false)
+                        }
+                    }
+                    "~" -> {
+                        if (rem.tail.head?.type == TokenType.OPERATOR && rem.tail.head?.keywordText == ">") {
+                            rem = rem.tail
+                            EdgeType(left = false, right = true, undirected = true)
+                        } else {
+                            EdgeType(left = false, right = false, undirected = true)
+                        }
+                    }
+                    else -> errorEdgeParse()
+                }
+            } else {
+                errorEdgeParse()
+            }
+            rem = rem.tail
+
+            return direction
+        }
+
+        fun parseEdgeWithSpec(): ParseNode {
+            val dir1 = parseLeftEdgePattern()
+            expect1(TokenType.LEFT_BRACKET, ErrorCode.PARSE_EXPECTED_LEFT_BRACKET_FOR_MATCH_EDGE, "match edge")
+            val name = parseName()
+            rem = name?.remaining ?: rem
+            val label = parseLabel()
+            rem = label?.remaining ?: rem
+            expect1(TokenType.RIGHT_BRACKET, ErrorCode.PARSE_EXPECTED_RIGHT_BRACKET_FOR_MATCH_EDGE, "match edge")
+            val dir2 = parseRightEdgePattern()
+
+            val dir = if (dir1 == EdgeType(left = false, right = true, undirected = false) &&
+                dir2 == EdgeType(left = true, right = false, undirected = false)
```

Comment:
Nit.: If not mistaken this is for _left or right (`<->`)_, in that case this could also be a function on `EdgeType` to detect?